shell/imports/client/grain/grainview.js

```diff
@@ -779,6 +780,30 @@ class GrainView {
     this._dep.changed();
   }
 
+  setShareData(shareData) {
+    var result = {};
+    if ("hash" in shareData) {
+      if (Match.test(shareData.hash, String)) {
```

Comment:
Thanks for bringing this up.  Since `setShareData` is only called as a result of an app's `postMessage`, I feel like it should give some meaningful feedback, thus the `console.warn` calls.  Are you suggesting that we do something like this instead?
```javascript
try {
  check(shareData, {"hash": Match.Maybe(String), "pathname": Match.Maybe(String)});
catch (err) {
  console.error(err);
  throw err;
}
```
A silent failure or a unhelpful error on the console will not make Sandstorm more usable.  I'm sure you will agree.

Reply:
That's roughly what I had in mind. As I see it there are two considerations:

1. For usability, app authors should get a clear indication of what is wrong.
2. Since this code is security sensitive (validating untrusted input), we want it to be as clear as possible when reading the implementation that it is correct.

From the perspective of (1), we want:

- The error to be loud -- with the current patch, if the app author feeds the API a bogus argument, then the sharing link will just be wrong. Especially since the window size sometimes cuts off the URL anyway, this is too subtle -- even though there is at least a warning at the console, it might be easy to miss.
- The error should be intelligible. I might add some explanatory text to the call to console.error, rather than expecting an app developer to decipher the exception & stack trace alone.

From the perspective of (2), just having to look at a call to check() makes it easier to understand what the shape of the data is vs reading a chunk of imperative checks. It would actually be better in this regard to move the check into the place where the postMessage() is received; this way it's clear that validation is the first thing that happens (this would also be consistent with how the rest of the postMessage APIs do things).

So I guess I'm making two suggestions:

1. Change this to just a `try { check(...) } catch (...) { console.error(...); throw ... }` per above, with some explanatory text logged in addition to the raw error, and
2. Move this logic to the `postMessage` handler, so it's clear looking at the point of receipt that this is being validated.

Reply:
That works for me.  I like moving it to the postMessage handling.  I've filed #3464 to remind us/me to refactor the huge postMessage handling function.

Reply:
I believe that this is resolved in https://github.com/sandstorm-io/sandstorm/pull/3463/commits/f0cb5dd2a84d504907abadc0677a423287d5cafd.